api/external.go

```diff
@@ -331,6 +207,150 @@ func (a *API) internalExternalProviderCallback(w http.ResponseWriter, r *http.Re
+			}
+			if user != nil {
+				if identity, terr = a.createNewIdentity(tx, user, providerType, identityData); terr != nil {
+					return nil, terr
+				}
+				if terr = user.UpdateAppMetaDataProviders(tx); terr != nil {
+					return nil, terr
+				}
+			} else {
+				if config.DisableSignup {
+					return nil, forbiddenError("Signups not allowed for this instance")
+				}
+
+				// prefer primary email for new signups
+				emailData = userData.Emails[0]
+				for _, e := range userData.Emails {
+					if e.Primary {
+						emailData = e
+						break
+					}
+				}
+
+				params := &SignupParams{
+					Provider: providerType,
+					Email:    emailData.Email,
+					Aud:      aud,
+					Data:     identityData,
+				}
+
+				user, terr = a.signupNewUser(ctx, tx, params)
+				if terr != nil {
+					return nil, terr
+				}
+
+				if identity, terr = a.createNewIdentity(tx, user, providerType, identityData); terr != nil {
+					return nil, terr
+				}
+			}
+		} else {
+			return nil, terr
+		}
+	}
+
+	if identity != nil && user == nil {
+		// get user associated with identity
+		user, terr = models.FindUserByID(tx, identity.UserID)
+		if terr != nil {
+			return nil, terr
+		}
+		identity.IdentityData = identityData
+		if terr = tx.UpdateOnly(identity, "identity_data", "last_sign_in_at"); terr != nil {
+			return nil, terr
+		}
+		// email & verified status might have changed if identity's email changed
+		emailData = provider.Email{
+			Email:    userData.Metadata.Email,
+			Verified: userData.Metadata.EmailVerified,
+		}
+		if terr = user.UpdateUserMetaData(tx, identityData); terr != nil {
+			return nil, terr
+		}
+		if terr = user.UpdateAppMetaDataProviders(tx); terr != nil {
+			return nil, terr
+		}
+	}
+
+	if user.IsBanned() {
+		return nil, unauthorizedError("User is unauthorized")
+	}
+
+	// an account with a previously unconfirmed email + password
+	// combination or phone may exist. so now that there is an
+	// OAuth identity bound to this user, and since they have not
+	// confirmed their email or phone, they are unaware that a
+	// potentially malicious door exists into their account; thus
+	// the password and phone needs to be removed.
+	if terr = user.RemoveUnconfirmedIdentities(tx); terr != nil {
+		return nil, internalServerError("Error updating user").WithInternalError(terr)
+	}
+
+	if !user.IsConfirmed() {
+		if !emailData.Verified && !config.Mailer.Autoconfirm {
+			mailer := a.Mailer(ctx)
+			referrer := a.getReferrer(r)
+			if terr = sendConfirmation(tx, user, mailer, config.SMTP.MaxFrequency, referrer, config.Mailer.OtpLength); terr != nil {
+				if errors.Is(terr, MaxFrequencyLimitError) {
+					return nil, tooManyRequestsError("For security purposes, you can only request this once every minute")
+				}
+				return nil, internalServerError("Error sending confirmation mail").WithInternalError(terr)
+			}
+			// email must be verified to issue a token
+			return nil, errReturnNil
```

Comment:
since this is the only place where we return `errReturnNil`, what do you think of making the error more obvious like `errEmailNotVerified` ?

Reply:
I'd rather fix this in another PR, want to keep this one fairly mechanical, just shuffling things around not introducing new behavior.